// mcp-auth/lib/auth/authorization-server-handler.js
import cors from 'cors';
import { Router } from 'express';
import snakecaseKeys from 'snakecase-keys';
import { serverMetadataPaths } from '../utils/fetch-server-config.js';
import { validateAuthServer } from '../utils/validate-auth-server.js';
import { AuthServerMetadataCache } from './auth-server-metadata-cache.js';
import { MCPAuthHandler } from './mcp-auth-handler.js';
import { TokenVerifier } from './token-verifier.js';
/**
 * Handles the authentication logic for the legacy `server` mode.
 * @deprecated
 */
export class AuthorizationServerHandler extends MCPAuthHandler {
    constructor(config) {
        super();
        this.config = config;
        /** Cache for auth server metadata, supporting discovery configs. */
        this.metadataCache = new AuthServerMetadataCache();
        console.warn('The authorization server mode is deprecated. Please use resource server mode instead.');
        validateAuthServer(config.server);
        this.tokenVerifier = new TokenVerifier([config.server]);
    }
    createMetadataRouter() {
        // eslint-disable-next-line new-cap
        const router = Router();
        router.use(serverMetadataPaths.oauth, cors());
        router.get(serverMetadataPaths.oauth, async (_, response) => {
            const metadata = await this.metadataCache.getMetadata(this.config.server);
            response.status(200).json(snakecaseKeys(metadata));
        });
        return router;
    }
    /**
     * This is a dummy implementation that ignores the options, as there is only
     * one `TokenVerifier` in the authorization server mode.
     */
    getTokenVerifier(_options) {
        return this.tokenVerifier;
    }
}
