// mcp-auth/lib/handlers/handle-bearer-auth.js
import { cond, condObject, trySafe } from '@silverhand/essentials';
import snakecaseKeys from 'snakecase-keys';
import { MCPAuthAuthServerError, MCPAuthBearerAuthError, MCPAuthConfigError, MCPAuthTokenVerificationError, } from '../errors.js';
import { BearerWWWAuthenticateHeader } from '../utils/bearer-www-authenticate-header.js';
import { createResourceMetadataEndpoint } from '../utils/create-resource-metadata-endpoint.js';
const getBearerTokenFromHeaders = (headers) => {
    const authHeader = headers.authorization;
    if (!authHeader) {
        throw new MCPAuthBearerAuthError('missing_auth_header');
    }
    const [scheme, token, ...rest] = authHeader.split(' ');
    if (scheme?.toLowerCase() !== 'bearer' || rest.length > 0) {
        throw new MCPAuthBearerAuthError('invalid_auth_header_format');
    }
    if (!token) {
        throw new MCPAuthBearerAuthError('missing_bearer_token');
    }
    return token;
};
const handleError = (error, response, resourceMetadataEndpoint, showErrorDetails = false) => {
    const wwwAuthenticateHeader = new BearerWWWAuthenticateHeader();
    if (error instanceof MCPAuthTokenVerificationError || error instanceof MCPAuthBearerAuthError) {
        wwwAuthenticateHeader.setParameterIfValueExists('error', error.code);
        wwwAuthenticateHeader.setParameterIfValueExists('error_description', error.message);
    }
    if (error instanceof MCPAuthTokenVerificationError) {
        wwwAuthenticateHeader.setParameterIfValueExists('resource_metadata', resourceMetadataEndpoint);
        response
            .set(wwwAuthenticateHeader.headerName, wwwAuthenticateHeader.toString())
            .status(401)
            .json(snakecaseKeys(error.toJson(showErrorDetails)));
        return;
    }
    if (error instanceof MCPAuthBearerAuthError) {
        const statusCode = error.code === 'missing_required_scopes' ? 403 : 401;
        if (statusCode === 401) {
            wwwAuthenticateHeader.setParameterIfValueExists('resource_metadata', resourceMetadataEndpoint);
        }
        response
            .set(wwwAuthenticateHeader.headerName, wwwAuthenticateHeader.toString())
            .status(statusCode)
            .json(snakecaseKeys(error.toJson(showErrorDetails)));
        return;
    }
    if (error instanceof MCPAuthAuthServerError || error instanceof MCPAuthConfigError) {
        response.status(500).json(snakecaseKeys(condObject({
            error: 'server_error',
            error_description: 'An error occurred with the authorization server.',
            cause: showErrorDetails ? error.toJson() : undefined,
        })));
        return;
    }
    throw error;
};
/**
 * Creates a middleware function for handling Bearer auth in an Express application.
 *
 * This middleware extracts the Bearer token from the `Authorization` header, verifies it using the
 * provided `verifyAccessToken` function, and checks the issuer, audience, and required scopes.
 *
 * - If the token is valid, it adds the auth information to the `request.auth` property;
 * if not, it responds with an appropriate error message.
 * - If access token verification fails, it responds with a 401 Unauthorized error.
 * - If the token does not have the required scopes, it responds with a 403 Forbidden error.
 * - If unexpected errors occur during the auth process, the middleware will re-throw them.
 *
 * **Note:**  The `request.auth` object will contain extended fields compared to the standard
 * {@link AuthInfo} interface defined in the `@modelcontextprotocol/sdk` module. See the extended
 * interface in this file for details.
 *
 * @param param0 Configuration for the Bearer auth handler.
 * @returns A middleware function for Express that handles Bearer auth.
 * @see {@link BearerAuthConfig} for the configuration options.
 */
export const handleBearerAuth = ({ verifyAccessToken, issuer, requiredScopes, audience, resource, showErrorDetails, }) => {
    if (typeof verifyAccessToken !== 'function') {
        throw new TypeError('`verifyAccessToken` must be a function that takes a token and returns an `AuthInfo` object.');
    }
    if (typeof issuer !== 'function' && typeof issuer !== 'string') {
        throw new TypeError('`issuer` must be either a string or a function that validates the token issuer.');
    }
    if (typeof issuer === 'string' && !trySafe(() => new URL(issuer))) {
        throw new TypeError(`\`issuer\` must be a valid URL.`);
    }
    const bearerAuthHandler = async function (request, response, next) {
        try {
            const token = getBearerTokenFromHeaders(request.headers);
            const authInfo = await verifyAccessToken(token);
            if (typeof issuer === 'function') {
                issuer(authInfo.issuer);
            }
            else {
                /**
                 * When issuer is provided as a string, directly compare it with the token's issuer value
                 */
                // eslint-disable-next-line no-lonely-if
                if (authInfo.issuer !== issuer) {
                    throw new MCPAuthBearerAuthError('invalid_issuer', {
                        expected: issuer,
                        actual: authInfo.issuer,
                    });
                }
            }
            if (audience &&
                (Array.isArray(authInfo.audience)
                    ? !authInfo.audience.includes(audience)
                    : authInfo.audience !== audience)) {
                throw new MCPAuthBearerAuthError('invalid_audience', {
                    expected: audience,
                    actual: authInfo.audience,
                });
            }
            if (requiredScopes) {
                const missingScopes = requiredScopes.filter((scope) => !authInfo.scopes.includes(scope));
                if (missingScopes.length > 0) {
                    throw new MCPAuthBearerAuthError('missing_required_scopes', {
                        missingScopes,
                    });
                }
            }
            if (request.auth) {
                console.warn('Request already contains auth info and will be overwritten. Please double-check if this is intended.');
            }
            // eslint-disable-next-line @silverhand/fp/no-mutation
            request.auth = authInfo;
            next();
        }
        catch (error) {
            handleError(error, response, cond(resource && createResourceMetadataEndpoint(resource).toString()), showErrorDetails);
        }
    };
    return bearerAuthHandler;
};
