// @silverhand/essentials/lib/index.cjs
'use strict';

var array = require('./utilities/array.cjs');
var assert = require('./utilities/assert.cjs');
var assertions = require('./utilities/assertions.cjs');
var conditional = require('./utilities/conditional.cjs');
var env = require('./utilities/env.cjs');
var _function = require('./utilities/function.cjs');
var pick = require('./utilities/pick.cjs');
var string = require('./utilities/string.cjs');
var get = require('./utilities/get.cjs');
var url = require('./utilities/url.cjs');
var object = require('./utilities/object.cjs');
var once = require('./utilities/once.cjs');



exports.deduplicate = array.deduplicate;
exports.deduplicateByKey = array.deduplicateByKey;
exports.normalizeValueToStringArray = array.normalizeValueToStringArray;
exports.repeat = array.repeat;
exports.assert = assert.assert;
exports.has = assertions.has;
exports.hasChineseCharacter = assertions.hasChineseCharacter;
exports.isPlainObject = assertions.isPlainObject;
exports.isSameArray = assertions.isSameArray;
exports.notEmpty = assertions.notEmpty;
exports.notFalsy = assertions.notFalsy;
exports.cond = conditional.cond;
exports.condArray = conditional.condArray;
exports.condObject = conditional.condObject;
exports.condString = conditional.condString;
exports.conditional = conditional.conditional;
exports.conditionalArray = conditional.conditionalArray;
exports.conditionalObject = conditional.conditionalObject;
exports.conditionalString = conditional.conditionalString;
exports.nullable = conditional.nullable;
exports.assertEnv = env.assertEnv;
exports.getEnv = env.getEnv;
exports.getEnvAsStringArray = env.getEnvAsStringArray;
exports.isNode = env.isNode;
exports.isPromise = _function.isPromise;
exports.noop = _function.noop;
exports.trySafe = _function.trySafe;
exports.tryThat = _function.tryThat;
exports.pick = pick.pick;
exports.pickState = pick.pickState;
exports.toTitle = string.toTitle;
exports.urlSafeBase64 = string.urlSafeBase64;
exports.yes = string.yes;
exports.get = get.get;
exports.getSafe = get.getSafe;
exports.getValue = get.getValue;
exports.appendPath = url.appendPath;
exports.joinPath = url.joinPath;
exports.isKeyInObject = object.isKeyInObject;
exports.isObject = object.isObject;
exports.removeUndefinedKeys = object.removeUndefinedKeys;
exports.once = once.once;
