// mcp-auth/lib/errors.js
import { cond, condObject } from '@silverhand/essentials';
/**
 * Base class for all mcp-auth errors.
 *
 * It provides a standardized way to handle errors related to MCP authentication and authorization.
 */
export class MCPAuthError extends Error {
    constructor(
    /**
     * The error code in snake_case format.
     */
    code, 
    /**
     * A human-readable description of the error.
     */
    message) {
        super(message);
        this.code = code;
        this.name = 'MCPAuthError';
    }
    /**
     * Converts the error to a HTTP response friendly JSON format.
     *
     * @param showCause Whether to include the cause of the error in the JSON response.
     * Defaults to `false`.
     */
    toJson(showCause = false) {
        return condObject({
            error: this.code,
            errorDescription: this.message,
            cause: cond(showCause && this.cause),
        });
    }
}
/**
 * Error thrown when there is a configuration issue with mcp-auth.
 */
export class MCPAuthConfigError extends MCPAuthError {
    constructor() {
        super(...arguments);
        this.name = 'MCPAuthConfigError';
    }
}
export const authServerErrorDescription = Object.freeze({
    invalid_server_metadata: 'The server metadata is invalid or malformed.',
    invalid_server_config: 'The server configuration does not match the MCP specification.',
    missing_jwks_uri: 'The server metadata does not contain a JWKS URI, which is required for JWT verification.',
});
/**
 * Error thrown when there is an issue with the remote authorization server.
 */
export class MCPAuthAuthServerError extends MCPAuthError {
    constructor(code, cause) {
        super(code, authServerErrorDescription[code] || 'An error occurred with the authorization server.');
        this.code = code;
        this.cause = cause;
        this.name = 'MCPAuthAuthServerError';
    }
}
export const bearerAuthErrorDescription = Object.freeze({
    missing_auth_header: 'Missing `Authorization` header. Please provide a valid bearer token.',
    invalid_auth_header_format: 'Invalid `Authorization` header format. Expected "Bearer <token>".',
    missing_bearer_token: 'Missing bearer token in `Authorization` header. Please provide a valid token.',
    invalid_issuer: 'The token issuer does not match the expected issuer.',
    invalid_audience: 'The token audience does not match the expected audience.',
    missing_required_scopes: 'The token does not contain the necessary scopes for this request.',
    invalid_token: 'The provided token is not valid or has expired.',
});
/**
 * Error thrown when there is an issue when authenticating with Bearer tokens.
 */
export class MCPAuthBearerAuthError extends MCPAuthError {
    constructor(code, cause) {
        super(code, bearerAuthErrorDescription[code] || 'An error occurred with the Bearer auth.');
        this.code = code;
        this.cause = cause;
        this.name = 'MCPAuthBearerAuthError';
    }
    toJson(showCause = false) {
        // Matches the OAuth 2.0 error response format at best effort
        return condObject({
            ...super.toJson(showCause),
            errorUri: this.cause?.uri?.href,
            missingScopes: this.cause?.missingScopes,
        });
    }
}
export const tokenVerificationErrorDescription = Object.freeze({
    invalid_token: 'The provided token is invalid or malformed.',
    token_verification_failed: 'Token verification failed due to an error.',
});
/**
 * Error thrown when there is an issue when verifying tokens.
 */
export class MCPAuthTokenVerificationError extends MCPAuthError {
    constructor(code, cause) {
        super(code, tokenVerificationErrorDescription[code] || 'An error occurred while verifying the token.');
        this.code = code;
        this.cause = cause;
        this.name = 'MCPAuthTokenVerificationError';
    }
}
