// mcp-auth/lib/index.js
import { AuthorizationServerHandler, } from './auth/authorization-server-handler.js';
import { ResourceServerHandler, } from './auth/resource-server-handler.js';
import { MCPAuthAuthServerError } from './errors.js';
import { handleBearerAuth, } from './handlers/handle-bearer-auth.js';
export * from './types/oauth.js';
export * from './types/auth-server.js';
export * from './errors.js';
export * from './handlers/handle-bearer-auth.js';
export * from './utils/fetch-server-config.js';
export * from './utils/validate-server-config.js';
export * from './utils/create-verify-jwt.js';
/**
 * The main class for the mcp-auth library. It acts as a factory and registry for creating
 * authentication policies for your protected resources.
 *
 * It is initialized with your server configurations and provides a `bearerAuth` method
 * to generate Express middleware for token-based authentication.
 *
 * @example
 * ### Usage in `resource server` mode
 *
 * This is the recommended approach for new applications.
 *
 * #### Option 1: Discovery config (recommended for edge runtimes)
 *
 * Use this when you want metadata to be fetched on-demand. This is especially useful for
 * edge runtimes like Cloudflare Workers where top-level async fetch is not allowed.
 *
 * ```ts
 * import express from 'express';
 * import { MCPAuth } from 'mcp-auth';
 *
 * const app = express();
 * const resourceIdentifier = 'https://api.example.com/notes';
 *
 * const mcpAuth = new MCPAuth({
 *   protectedResources: [
 *     {
 *       metadata: {
 *         resource: resourceIdentifier,
 *         // Just pass issuer and type - metadata will be fetched on first request
 *         authorizationServers: [{ issuer: 'https://auth.logto.io/oidc', type: 'oidc' }],
 *         scopesSupported: ['read:notes', 'write:notes'],
 *       },
 *     },
 *   ],
 * });
 * ```
 *
 * #### Option 2: Resolved config (pre-fetched metadata)
 *
 * Use this when you want to fetch and validate metadata at startup time.
 *
 * ```ts
 * import express from 'express';
 * import { MCPAuth, fetchServerConfig } from 'mcp-auth';
 *
 * const app = express();
 * const resourceIdentifier = 'https://api.example.com/notes';
 * const authServerConfig = await fetchServerConfig('https://auth.logto.io/oidc', { type: 'oidc' });
 *
 * const mcpAuth = new MCPAuth({
 *   protectedResources: [
 *     {
 *       metadata: {
 *         resource: resourceIdentifier,
 *         authorizationServers: [authServerConfig],
 *         scopesSupported: ['read:notes', 'write:notes'],
 *       },
 *     },
 *   ],
 * });
 * ```
 *
 * #### Using the middleware
 *
 * ```ts
 * // Mount the router to handle Protected Resource Metadata
 * app.use(mcpAuth.protectedResourceMetadataRouter());
 *
 * // Protect an API endpoint for the configured resource
 * app.get(
 *   '/notes',
 *   mcpAuth.bearerAuth('jwt', {
 *     resource: resourceIdentifier, // Specify which resource this endpoint belongs to
 *     audience: resourceIdentifier, // Optionally, validate the 'aud' claim
 *     requiredScopes: ['read:notes'],
 *   }),
 *   (req, res) => {
 *     console.log('Auth info:', req.auth);
 *     res.json({ notes: [] });
 *   },
 * );
 * ```
 *
 * ### Legacy Usage in `authorization server` mode (Deprecated)
 *
 * This approach is supported for backward compatibility.
 *
 * ```ts
 * import express from 'express';
 * import { MCPAuth } from 'mcp-auth';
 *
 * const app = express();
 * const mcpAuth = new MCPAuth({
 *   // Discovery config - metadata fetched on-demand
 *   server: { issuer: 'https://auth.logto.io/oidc', type: 'oidc' },
 * });
 *
 * // Mount the router to handle legacy Authorization Server Metadata
 * app.use(mcpAuth.delegatedRouter());
 *
 * // Protect an endpoint using the default policy
 * app.get(
 *   '/mcp',
 *   mcpAuth.bearerAuth('jwt', { requiredScopes: ['read', 'write'] }),
 *   (req, res) => {
 *     console.log('Auth info:', req.auth);
 *     // Handle the MCP request here
 *   },
 * );
 * ```
 */
export class MCPAuth {
    /**
     * Creates an instance of MCPAuth.
     * It validates the entire configuration upfront to fail fast on errors.
     * @param config The authentication configuration.
     */
    constructor(config) {
        this.config = config;
        if ('server' in config) {
            this.authHandler = new AuthorizationServerHandler(config);
            return;
        }
        if (!('protectedResources' in config)) {
            throw new MCPAuthAuthServerError('invalid_server_config', {
                cause: 'No authorization server or protected resource metadata is provided.',
            });
        }
        this.authHandler = new ResourceServerHandler(config);
    }
    /**
     * Creates a delegated router for serving legacy OAuth 2.0 Authorization Server Metadata endpoint
     * (`/.well-known/oauth-authorization-server`) with the metadata provided to the instance.
     *
     * @deprecated Use {@link protectedResourceMetadataRouter} instead.
     *
     * @example
     * ```ts
     * import express from 'express';
     * import { MCPAuth } from 'mcp-auth';
     *
     * const app = express();
     * const mcpAuth: MCPAuth; // Assume this is initialized
     * app.use(mcpAuth.delegatedRouter());
     * ```
     *
     * @returns A router that serves the OAuth 2.0 Authorization Server Metadata endpoint with the
     * metadata provided to the instance.
     * @throws {MCPAuthAuthServerError} If called in `resource server` mode.
     */
    delegatedRouter() {
        if (this.authHandler instanceof ResourceServerHandler) {
            throw new MCPAuthAuthServerError('invalid_server_config', {
                cause: '`delegatedRouter` is not available in `resource server` mode.',
            });
        }
        return this.authHandler.createMetadataRouter();
    }
    /**
     * Creates a router that serves the OAuth 2.0 Protected Resource Metadata endpoint
     * for all configured resources.
     *
     * This router automatically creates the correct `.well-known` endpoints for each
     * resource identifier provided in your configuration.
     *
     * @returns A router that serves the OAuth 2.0 Protected Resource Metadata endpoint.
     * @throws {MCPAuthAuthServerError} If called in `authorization server` mode.
     *
     * @example
     * ```ts
     * import express from 'express';
     * import { MCPAuth } from 'mcp-auth';
     *
     * // Assuming mcpAuth is initialized with one or more `protectedResources` configs
     * const mcpAuth: MCPAuth;
     * const app = express();
     *
     * // This will serve metadata at `/.well-known/oauth-protected-resource/...`
     * // based on your resource identifiers.
     * app.use(mcpAuth.protectedResourceMetadataRouter());
     * ```
     */
    protectedResourceMetadataRouter() {
        if (this.authHandler instanceof AuthorizationServerHandler) {
            throw new MCPAuthAuthServerError('invalid_server_config', {
                cause: '`protectedResourceMetadataRouter` is not available in `authorization server` mode.',
            });
        }
        return this.authHandler.createMetadataRouter();
    }
    bearerAuth(modeOrVerify, { jwtVerify, remoteJwkSet, ...config } = {}) {
        /**
         * The `resource` property in the config is crucial for selecting the correct TokenVerifier
         * in resource server mode. This check ensures it's not forgotten.
         */
        if ('protectedResources' in this.config && !config.resource) {
            throw new MCPAuthAuthServerError('invalid_server_config', {
                cause: 'A `resource` must be specified in the `bearerAuth` configuration when using a `protectedResources` configuration.',
            });
        }
        /**
         * In the deprecated `authorization server` mode, `getTokenVerifier` does not utilize the
         * `resource` parameter. Passing an empty string `''` is a straightforward approach that
         * avoids over-engineering a solution for a legacy path.
         */
        const tokenVerifier = this.authHandler.getTokenVerifier({ resource: config.resource ?? '' });
        const getVerifyFunction = () => {
            if (typeof modeOrVerify === 'function') {
                return modeOrVerify;
            }
            // If a mode is provided, create the verification function from the TokenVerifier.
            switch (modeOrVerify) {
                case 'jwt': {
                    return tokenVerifier.createVerifyJwtFunction({ jwtVerify, remoteJwkSet });
                }
            }
        };
        return handleBearerAuth({
            verifyAccessToken: getVerifyFunction(),
            issuer: tokenVerifier.getJwtIssuerValidator(),
            ...config,
        });
    }
}
