// mcp-auth/lib/utils/validate-server-config.js
import { condObject } from '@silverhand/essentials';
import { camelCaseAuthorizationServerMetadataSchema, defaultValues } from '../types/oauth.js';
const authServerSuccessDescription = Object.freeze({
    server_metadata_valid: 'The server metadata is valid and conforms to the expected schema.',
    dynamic_registration_supported: 'Dynamic Client Registration (RFC 7591) is supported by the server.',
    pkce_supported: 'Proof Key for Code Exchange (PKCE) is supported by the server.',
    s256_code_challenge_method_supported: 'The "S256" code challenge method for Proof Key for Code Exchange (PKCE) is supported by the server.',
    authorization_code_grant_supported: 'The "authorization_code" grant type is supported by the server.',
    code_response_type_supported: 'The "code" response type is supported by the server.',
});
const createSuccess = (code) => ({
    code,
    description: authServerSuccessDescription[code],
});
const authServerConfigErrorDescription = Object.freeze({
    invalid_server_metadata: 'The server metadata is not a valid object or does not conform to the expected schema.',
    code_response_type_not_supported: 'The server does not support the "code" response type or the "code" response type is not included in one of the supported response types.',
    authorization_code_grant_not_supported: 'The server does not support the "authorization_code" grant type.',
    pkce_not_supported: 'The server does not support Proof Key for Code Exchange (PKCE).',
    s256_code_challenge_method_not_supported: 'The server does not support the "S256" code challenge method for Proof Key for Code Exchange (PKCE).',
});
const createError = (code, cause) => ({
    code,
    description: authServerConfigErrorDescription[code],
    cause,
});
const authServerConfigWarningDescription = Object.freeze({
    dynamic_registration_not_supported: 'Dynamic Client Registration (RFC 7591) is not supported by the server.',
});
const createWarning = (code) => ({
    code,
    description: authServerConfigWarningDescription[code],
});
// eslint-disable-next-line complexity
export const validateServerConfig = (config, verbose = false) => {
    const { metadata } = config;
    const errors = [];
    const warnings = [];
    const successes = [];
    const parsed = camelCaseAuthorizationServerMetadataSchema.safeParse(metadata);
    const buildReturnValue = () => 
    // eslint-disable-next-line no-restricted-syntax -- the return type inference is beyond TypeScript's capabilities
    condObject({
        isValid: errors.length === 0,
        errors: errors.length > 0 ? errors : undefined,
        warnings,
        successes: verbose ? successes : undefined,
    });
    /* eslint-disable @silverhand/fp/no-mutating-methods -- for the sake of readability */
    if (!parsed.success) {
        errors.push(createError('invalid_server_metadata', parsed.error));
        return buildReturnValue();
    }
    if (verbose) {
        successes.push(createSuccess('server_metadata_valid'));
    }
    if (!metadata.responseTypesSupported.some((type) => type.split(' ').includes('code'))) {
        errors.push(createError('code_response_type_not_supported'));
    }
    else if (verbose) {
        successes.push(createSuccess('code_response_type_supported'));
    }
    if (!(metadata.grantTypesSupported ?? defaultValues.grantTypesSupported)?.includes('authorization_code')) {
        errors.push(createError('authorization_code_grant_not_supported'));
    }
    else if (verbose) {
        successes.push(createSuccess('authorization_code_grant_supported'));
    }
    if (metadata.codeChallengeMethodsSupported) {
        if (verbose) {
            successes.push(createSuccess('pkce_supported'));
        }
        if (!metadata.codeChallengeMethodsSupported.includes('S256')) {
            errors.push(createError('s256_code_challenge_method_not_supported'));
        }
        else if (verbose) {
            successes.push(createSuccess('s256_code_challenge_method_supported'));
        }
    }
    else {
        errors.push(createError('pkce_not_supported'));
    }
    if (!metadata.registrationEndpoint) {
        warnings.push(createWarning('dynamic_registration_not_supported'));
    }
    else if (verbose) {
        successes.push(createSuccess('dynamic_registration_supported'));
    }
    /* eslint-enable @silverhand/fp/no-mutating-methods */
    return buildReturnValue();
};
